File: backend/upload-all-movies.ts
import { S3Client, PutObjectCommand } from '@aws-sdk/client-s3';
import { createClient } from '@supabase/supabase-js';
import * as fs from 'fs';
import * as path from 'path';
import * as dotenv from 'dotenv';

dotenv.config();

const s3Client = new S3Client({
  region: process.env.AWS_REGION || 'us-east-1',
  credentials: {
    accessKeyId: process.env.AWS_ACCESS_KEY_ID!,
    secretAccessKey: process.env.AWS_SECRET_ACCESS_KEY!,
  },
});

const supabase = createClient(
  process.env.SUPABASE_URL!,
  process.env.SUPABASE_SERVICE_ROLE_KEY || process.env.SUPABASE_ANON_KEY!
);

interface MovieMapping {
  folderName: string;
  dbTitle: string;
  contentId?: string;
}

const movieMappings: MovieMapping[] = [
  { folderName: 'FILME_ A Hora do Mal (2025)', dbTitle: 'A Hora do Mal' },
  { folderName: 'FILME_ A Longa Marcha - Caminhe ou Morra (2025)', dbTitle: 'A Longa Marcha - Caminhe ou Morra' },
  { folderName: 'FILME_ Como Treinar o Seu Dragão (2025)', dbTitle: 'Como Treinar o Seu Dragão' },
  { folderName: 'FILME_ Demon Slayer - Castelo Infinito (2025)', dbTitle: 'Demon Slayer - Castelo Infinito' },
  { folderName: 'FILME_ F1 - O Filme (2025)', dbTitle: 'F1 - O Filme' },
  { folderName: 'FILME_ Invocação do Mal 4_ O Último Ritual (2025)', dbTitle: 'Invocação do Mal 4: O Último Ritual' },
  { folderName: 'FILME_ Jurassic World_ Recomeço (2025)', dbTitle: 'Jurassic World: Recomeço' },
  { folderName: 'FILME_ Quarteto Fantástico 4 - Primeiros Passos (2025)', dbTitle: 'Quarteto Fantástico 4 - Primeiros Passos' },
];

async function uploadVideoToS3(filePath: string, s3Key: string): Promise<string> {
  console.log(`📤 Uploading ${path.basename(filePath)} to S3...`);

  const fileStream = fs.createReadStream(filePath);
  const stats = fs.statSync(filePath);

  const uploadParams = {
    Bucket: process.env.S3_VIDEO_BUCKET!,
    Key: s3Key,
    Body: fileStream,
    ContentType: 'video/x-matroska',
  };

  await s3Client.send(new PutObjectCommand(uploadParams));

  const s3Url = `https://${process.env.S3_VIDEO_BUCKET}.s3.${process.env.AWS_REGION}.amazonaws.com/${s3Key}`;
  console.log(`✅ Uploaded: ${s3Url}`);

  return s3Url;
}

async function updateContentInDatabase(contentId: string, videoUrl: string) {
  const { error } = await supabase
    .from('content')
    .update({
      video_url: videoUrl,
      processing_status: 'completed',
      updated_at: new Date().toISOString(),
    })
    .eq('id', contentId);

  if (error) {
    console.error(`❌ Error updating content ${contentId}:`, error);
    throw error;
  }

  console.log(`✅ Database updated for content ID: ${contentId}`);
}

async function processMovie(mapping: MovieMapping) {
  console.log(`\n${'='.repeat(60)}`);
  console.log(`🎬 Processing: ${mapping.dbTitle}`);
  console.log(`${'='.repeat(60)}`);

  // Get content ID from database
  const { data: content, error: fetchError } = await supabase
    .from('content')
    .select('id')
    .eq('title', mapping.dbTitle)
    .single();

  if (fetchError || !content) {
    console.error(`❌ Movie not found in database: ${mapping.dbTitle}`);
    return;
  }

  const contentId = content.id;
  console.log(`📋 Content ID: ${contentId}`);

  // Find video file
  const movieDir = path.join('E:', 'movies', mapping.folderName);

  if (!fs.existsSync(movieDir)) {
    console.error(`❌ Directory not found: ${movieDir}`);
    return;
  }

  const files = fs.readdirSync(movieDir);
  const videoFile = files.find(f => f.includes('DUBLADO') && (f.endsWith('.mkv') || f.endsWith('.mp4')));

  if (!videoFile) {
    console.error(`❌ Video file not found in: ${movieDir}`);
    return;
  }

  const videoPath = path.join(movieDir, videoFile);
  console.log(`📁 Found video: ${videoFile}`);

  const fileStats = fs.statSync(videoPath);
  console.log(`📊 File size: ${(fileStats.size / (1024 * 1024 * 1024)).toFixed(2)} GB`);

  // Generate S3 key
  const sanitizedTitle = mapping.dbTitle
    .toLowerCase()
    .normalize('NFD')
    .replace(/[\u0300-\u036f]/g, '')
    .replace(/[^a-z0-9]+/g, '-')
    .replace(/^-+|-+$/g, '');

  const ext = path.extname(videoFile);
  const s3Key = `videos/${sanitizedTitle}-2025-dubbed${ext}`;

  // Upload to S3
  const videoUrl = await uploadVideoToS3(videoPath, s3Key);

  // Update database
  await updateContentInDatabase(contentId, videoUrl);

  console.log(`✅ Successfully processed: ${mapping.dbTitle}\n`);
}

async function main() {
  console.log('🚀 Starting bulk movie upload process...\n');
  console.log(`📦 S3 Bucket: ${process.env.S3_VIDEO_BUCKET}`);
  console.log(`🌍 Region: ${process.env.AWS_REGION}`);
  console.log(`📊 Total movies to process: ${movieMappings.length}\n`);

  for (const mapping of movieMappings) {
    try {
      await processMovie(mapping);
    } catch (error) {
      console.error(`❌ Failed to process ${mapping.dbTitle}:`, error);
      console.log('⏭️  Continuing with next movie...\n');
    }
  }

  console.log('\n' + '='.repeat(60));
  console.log('🎉 Bulk upload process completed!');
  console.log('='.repeat(60));
}

main().catch(console.error);

File: upload-all-movies.py
# -*- coding: utf-8 -*-
import sys
import os

# Fix Windows console encoding
if sys.platform == 'win32':
    sys.stdout.reconfigure(encoding='utf-8')
    sys.stderr.reconfigure(encoding='utf-8')

import re
import boto3
import psycopg2
from pathlib import Path
from datetime import datetime

# Configuração AWS
AWS_REGION = 'us-east-2'
AWS_ACCESS_KEY_ID = os.getenv('AWS_ACCESS_KEY_ID', 'your_aws_access_key_id_here')
AWS_SECRET_ACCESS_KEY = os.getenv('AWS_SECRET_ACCESS_KEY', 'your_aws_secret_access_key_here')
S3_VIDEO_BUCKET = 'cinevision-video'
S3_COVER_BUCKET = 'cinevision-cover'

# Configuração Supabase
SUPABASE_DB = {
    'host': 'aws-1-sa-east-1.pooler.supabase.com',
    'port': 5432,
    'database': 'postgres',
    'user': 'postgres.szghyvnbmjlquznxhqum',
    'password': 'Umeomesmo1,'
}

# Diretório dos filmes
MOVIES_DIR = 'E:/movies'

# Preços oficiais dos filmes (em centavos) - DEFINIDOS NO README.md
OFFICIAL_PRICES = {
    'Quarteto Fantástico 4 - Primeiros Passos': 690,
    'Invocação do Mal 4_ O Último Ritual': 720,
    'Demon Slayer - Castelo Infinito': 750,
    'A Longa Marcha - Caminhe ou Morra': 700,
    'Superman': 710,
    'Como Treinar o Seu Dragão': 698,
    'Jurassic World_ Recomeço': 685,
    'F1 - O Filme': 705,
    'Lilo & Stitch': 698,
    'A Hora do Mal': 695,
}

def get_movie_price(title):
    """Retorna o preço oficial do filme em centavos"""
    # Normalizar título para busca
    normalized_title = title.strip()

    # Buscar preço exato
    if normalized_title in OFFICIAL_PRICES:
        price_cents = OFFICIAL_PRICES[normalized_title]
        print(f"  Price: R$ {price_cents/100:.2f} ({price_cents} centavos)")
        return price_cents

    # Buscar por similaridade parcial
    for official_title, price in OFFICIAL_PRICES.items():
        if official_title.lower() in normalized_title.lower() or normalized_title.lower() in official_title.lower():
            price_cents = price
            print(f"  Price: R$ {price_cents/100:.2f} ({price_cents} centavos)")
            return price_cents

    # Preço padrão caso não encontre
    print(f"  ⚠️  Price not found for '{title}', using default: R$ 7,00 (700 centavos)")
    return 700

def clean_title(folder_name):
    """Extrai título limpo do nome da pasta"""
    title = re.sub(r'^FILME_\s*', '', folder_name)
    title = re.sub(r'\s*\(\d{4}\)$', '', title)
    return title.strip()

def get_year(folder_name):
    """Extrai ano do nome da pasta"""
    match = re.search(r'\((\d{4})\)', folder_name)
    return int(match.group(1)) if match else 2025

def find_video_files(movie_dir):
    """Encontra arquivos de vídeo (DUBLADO e LEGENDADO)"""
    videos = {}
    for file in os.listdir(movie_dir):
        if file.endswith(('.mp4', '.mkv', '.avi')):
            if 'DUBLADO' in file.upper():
                videos['dubbed'] = os.path.join(movie_dir, file)
            elif 'LEGENDADO' in file.upper():
                videos['subtitled'] = os.path.join(movie_dir, file)
    return videos

def find_cover_image(movie_dir):
    """Encontra imagem de capa"""
    for file in os.listdir(movie_dir):
        if file.lower().endswith(('.jpg', '.jpeg', '.png', '.webp')):
            if any(keyword in file.lower() for keyword in ['capa', 'cover', 'poster']):
                return os.path.join(movie_dir, file)
    # Se não encontrar, pega a primeira imagem
    for file in os.listdir(movie_dir):
        if file.lower().endswith(('.jpg', '.jpeg', '.png', '.webp')):
            return os.path.join(movie_dir, file)
    return None

def upload_to_s3(file_path, bucket, key):
    """Faz upload de arquivo para S3"""
    print(f"  Uploading {os.path.basename(file_path)} to s3://{bucket}/{key}")
    s3_client = boto3.client(
        's3',
        region_name=AWS_REGION,
        aws_access_key_id=AWS_ACCESS_KEY_ID,
        aws_secret_access_key=AWS_SECRET_ACCESS_KEY
    )

    file_size = os.path.getsize(file_path)
    print(f"     Size: {file_size / (1024**3):.2f} GB")

    # Upload com progress callback
    def progress(bytes_transferred):
        percent = (bytes_transferred / file_size) * 100
        print(f"     Progress: {percent:.1f}%", end='\r')

    s3_client.upload_file(file_path, bucket, key, Callback=progress)
    print()  # Nova linha após o progress

    # Retorna URL pública
    url = f"https://{bucket}.s3.{AWS_REGION}.amazonaws.com/{key}"
    print(f"  ✅ Uploaded to: {url}")
    return url, key

def create_content_in_db(content_data):
    """Cria registro de conteúdo no banco de dados"""
    conn = psycopg2.connect(**SUPABASE_DB)
    cursor = conn.cursor()

    try:
        # Inserir conteúdo principal
        cursor.execute("""
            INSERT INTO content (
                title, description, release_year, content_type,
                poster_url, cover_storage_key, price_cents, currency,
                status, is_online, created_at, updated_at
            ) VALUES (
                %(title)s, %(description)s, %(release_year)s, 'movie',
                %(poster_url)s, %(cover_storage_key)s, %(price_cents)s, 'BRL',
                'ACTIVE', true, NOW(), NOW()
            ) RETURNING id
        """, content_data)

        content_id = cursor.fetchone()[0]
        print(f"  ✅ Created content record: {content_id}")

        conn.commit()
        return content_id
    except Exception as e:
        conn.rollback()
        print(f"  ❌ Error creating content: {e}")
        raise
    finally:
        cursor.close()
        conn.close()

def create_content_language(content_id, language_data):
    """Cria registro de idioma para o conteúdo"""
    conn = psycopg2.connect(**SUPABASE_DB)
    cursor = conn.cursor()

    try:
        cursor.execute("""
            INSERT INTO content_languages (
                content_id, language_type, language_code, language_name,
                video_url, video_storage_key, upload_status,
                is_active, is_default, created_at, updated_at
            ) VALUES (
                %(content_id)s, %(language_type)s, %(language_code)s, %(language_name)s,
                %(video_url)s, %(video_storage_key)s, 'completed',
                true, %(is_default)s, NOW(), NOW()
            ) RETURNING id
        """, language_data)

        language_id = cursor.fetchone()[0]
        conn.commit()
        print(f"  ✅ Created language record: {language_id} ({language_data['language_type']})")
        return language_id
    except Exception as e:
        conn.rollback()
        print(f"  ❌ Error creating language: {e}")
        raise
    finally:
        cursor.close()
        conn.close()

def create_purchase_for_user(user_email, content_id, price_cents):
    """Cria compra para um usuário"""
    conn = psycopg2.connect(**SUPABASE_DB)
    cursor = conn.cursor()

    try:
        # Buscar user_id
        cursor.execute("SELECT id FROM users WHERE email = %s", (user_email,))
        result = cursor.fetchone()

        if not result:
            print(f"  ⚠️  User {user_email} not found")
            return None

        user_id = result[0]

        # Criar purchase
        cursor.execute("""
            INSERT INTO purchases (
                user_id, content_id, amount_cents, currency,
                status, created_at, updated_at
            ) VALUES (
                %s, %s, %s, 'BRL',
                'COMPLETED', NOW(), NOW()
            ) RETURNING id
        """, (user_id, content_id, price_cents))

        purchase_id = cursor.fetchone()[0]
        conn.commit()
        print(f"  ✅ Created purchase for {user_email}: {purchase_id}")
        return purchase_id
    except Exception as e:
        conn.rollback()
        print(f"  ❌ Error creating purchase: {e}")
        raise
    finally:
        cursor.close()
        conn.close()

def process_movie(movie_folder):
    """Processa um filme completo"""
    print(f"\n{'='*80}")
    print(f"Processing: {movie_folder}")
    print(f"{'='*80}")

    movie_dir = os.path.join(MOVIES_DIR, movie_folder)

    # 1. Extrair informações
    title = clean_title(movie_folder)
    year = get_year(movie_folder)
    print(f"Title: {title}")
    print(f"Year: {year}")

    # 2. Encontrar arquivos
    videos = find_video_files(movie_dir)
    cover = find_cover_image(movie_dir)

    print(f"Videos found: {len(videos)}")
    for lang, path in videos.items():
        print(f"   - {lang}: {os.path.basename(path)}")

    if cover:
        print(f"Cover: {os.path.basename(cover)}")

    # 3. Obter preço oficial
    price_cents = get_movie_price(title)

    # 4. Upload capa
    cover_url = None
    cover_key = None
    if cover:
        timestamp = int(datetime.now().timestamp())
        cover_key = f"covers/movies/{title.lower().replace(' ', '-')}-{timestamp}{Path(cover).suffix}"
        cover_url, cover_key = upload_to_s3(cover, S3_COVER_BUCKET, cover_key)

    # 5. Criar registro de conteúdo
    content_data = {
        'title': title,
        'description': f'{title} ({year})',
        'release_year': year,
        'poster_url': cover_url,
        'cover_storage_key': cover_key,
        'price_cents': price_cents
    }

    content_id = create_content_in_db(content_data)

    # 5. Upload vídeos e criar registros de idiomas
    for lang_type, video_path in videos.items():
        timestamp = int(datetime.now().timestamp())
        video_key = f"videos/movies/{title.lower().replace(' ', '-')}-{lang_type}-{timestamp}{Path(video_path).suffix}"
        video_url, video_key = upload_to_s3(video_path, S3_VIDEO_BUCKET, video_key)

        language_data = {
            'content_id': content_id,
            'language_type': lang_type,
            'language_code': 'pt-BR' if lang_type == 'dubbed' else 'en-US',
            'language_name': 'Português (Brasil)' if lang_type == 'dubbed' else 'English (US)',
            'video_url': video_url,
            'video_storage_key': video_key,
            'is_default': lang_type == 'dubbed'
        }

        create_content_language(content_id, language_data)

    # 6. Criar compra para usuário
    create_purchase_for_user('cinevision@teste.com', content_id, price_cents)

    print(f"\n✅ Movie '{title}' processed successfully!")
    print(f"   Content ID: {content_id}")
    print(f"   Price: R$ {price_cents/100:.2f}")
    print(f"   Languages: {len(videos)}")

    return content_id

if __name__ == '__main__':
    if len(sys.argv) < 2:
        print("Usage: python upload-single-movie.py <movie_folder_name>")
        print("\nAvailable movies:")
        movies = [
            "FILME_  Lilo & Stitch (2025)",
            "FILME_ A Hora do Mal (2025)",
            "FILME_ A Longa Marcha - Caminhe ou Morra (2025)",
            "FILME_ Como Treinar o Seu Dragão (2025)",
            "FILME_ Demon Slayer - Castelo Infinito (2025)",
            "FILME_ F1 - O Filme (2025)",
            "FILME_ Invocação do Mal 4_ O Último Ritual (2025)",
            "FILME_ Jurassic World_ Recomeço (2025)",
            "FILME_ Quarteto Fantástico 4 - Primeiros Passos (2025)",
        ]
        for i, movie in enumerate(movies, 1):
            print(f"  {i}. {movie}")
        sys.exit(1)

    movie_folder = sys.argv[1]
    process_movie(movie_folder)
